src/llmcompressor/modifiers/quantization/calibration.py

```diff
@@ -0,0 +1,251 @@
+    observer.
+
+    :param module: torch.nn.Module
+    :param base_name: substring used to fetch the observer, scales, and zp
+    :param value: torch.Tensor to be passed to the observer
+
+    """
+    # If empty tensor, can't update zp/scale
+    # Case for MoEs
+    if value.numel() == 0:
+        return
+
+    call_observer(
+        module=module,
+        base_name=base_name,
+        value=value,
+    )
+
+
+def calibrate_input_hook():
+    """
+    Hook to calibrate input activations.
+    Will call the observers to update the scales/zp before applying
+    input QDQ in the module's forward pass.
+    """
+
+    def hook_fn(module: Module, inp):
+        inp = inp[0] if isinstance(inp, tuple) else inp
+        calibrate_activations(module, value=inp, base_name="input")
+
+    return hook_fn
+
+
+def calibrate_output_hook():
+    """
+    Hook to calibrate output activations.
+    Will call the observers to update the scales/zp before applying
+    output QDQ.
+    """
+
+    def hook_fn(module: Module, inp, output: torch.Tensor):
+        calibrate_activations(
+            module,
+            value=output,
+            base_name="output",
+        )
+        output = forward_quantize(
+            module=module,
+            value=output,
+            base_name="output",
+            args=module.quantization_scheme.output_activations,
+        )
+        return output
+
+    return hook_fn
+
+
+def calibrate_kv_cache_input_hook():
+    """
+    Hook to update inputs to attention layers when running
+    kv_cache quantization. Will update the passed in
+    kv_cache to singleton QuantizedKVParameterCache.
+    """
+
+    def hook_fn(module: Module, args, kwargs):
+        kv_cache = getattr(module, "kv_cache")
+        kwargs["past_key_value"] = kv_cache
+        kwargs["use_cache"] = False
+        return args, kwargs
+
+    return hook_fn
+
+
+def calibrate_kv_cache_output_hook():
+    """
+    Hook to update k_scale and v_scale parameters when running kv_cache quantization.
+    """
+
+    def hook_fn(module: Module, inpt, output: torch.Tensor):
+        kv_cache = getattr(module, "kv_cache")
+        update_parameter_data(module, kv_cache.k_scales[module.layer_idx], "k_scale")
+        update_parameter_data(module, kv_cache.v_scales[module.layer_idx], "v_scale")
+
+    return hook_fn
+
+
+def set_unset_kv_cache(module: Module):
+    """
+    Set or Unset singleton QuantizedKVParameterCache for each
+    attn module when running kv_cache quantization.
+    """
+    if not hasattr(module, "quantization_scheme"):
+        return
+
+    if is_kv_cache_quant_scheme(module.quantization_scheme):
+        output_args = module.quantization_scheme.output_activations
+        kv_cache = QuantizedKVParameterCache(output_args)
+        if hasattr(module, "kv_cache"):
+            delattr(module, "kv_cache")
+        else:
+            setattr(module, "kv_cache", kv_cache)
```

Comment:
if has then delete, 
if no have set?

why?

Reply:
The name of the function is `set_unset_kv_cache`

Reply:
I agree that this should likely be two separate functions, if needed at all

Reply:
this is needed for kv_cache quantization. we were previously fetching the cache through the args

Reply:
If by ^ you mean is resetting the cache, the only reason if you need to reset in the previous code is if you keep quantized [k|v]-tensors, which we dont do in the previous code bc we are reusing the singletons.

the code already works as is, not nit picky, but might be able to delete the delattr here for the future